jwt validation

abel296 opened the pull request:
Features:

- Provide the system with the possibility of validating jwt
- Jwt config: validation included paths, allowed scopes and jwks url config
- Jwt config can be set per domain or in common config

abel296 commented:
Hi @fabiocicerchia , I don't have access to see why the snyk test is not passing, can you give me access permit to check it out please?

fabiocicerchia commented:
Hi @abel296, thank you for the PR.
I'll try to sort out the accesses to the project, possibly make it available publicly, in the meantime here are the errors raised by snyk:
![](https://github.com/fabiocicerchia/go-proxy-cache/assets/11521/839e8a59-5315-4ced-9467-ba39cc81fb96)

fabiocicerchia commented on config/config.go:
```diff
@@ -217,6 +218,13 @@ func (c *Configuration) copyOverWithLog(overrides Log) {
 
+// --- JWT.
+func (c *Configuration) copyOverWithJwt(overrides Jwt) {
+	c.Jwt.Included_paths = utils.Coalesce(overrides.Included_paths, c.Jwt.Included_paths).([]string)
+	c.Jwt.Allowed_scopes = utils.Coalesce(overrides.Allowed_scopes, c.Jwt.Allowed_scopes).([]string)
+	c.Jwt.Jwks_url = utils.Coalesce(overrides.Jwks_url, c.Jwt.Jwks_url).(string)
```
I'd change the following to be consistent with the naming convention of all other variables:

```suggestion
	c.Jwt.IncludedPaths = utils.Coalesce(overrides.IncludedPaths, c.Jwt.IncludedPaths).([]string)
	c.Jwt.AllowedScopes = utils.Coalesce(overrides.AllowedScopes, c.Jwt.AllowedScopes).([]string)
	c.Jwt.JwksUrl = utils.Coalesce(overrides.JwksUrl, c.Jwt.JwksUrl).(string)
```

abel296 replied:
done!

fabiocicerchia commented on config/model.go:
```diff
@@ -162,6 +165,21 @@ type DomainSet struct {
 
+// Jwt - Defines the config for the jwt validation.
+type Jwt struct {
+	Included_paths []string	`yaml:"included_paths" envconfig:"INCLUDED_PATHS"`
+	Allowed_scopes []string	`yaml:"allowed_scopes" envconfig:"ALLOWED_SCOPES"`
+	Jwks_url       string	`yaml:"jwks_url"`
```
same as above:

```suggestion
	IncludedPaths []string	`yaml:"included_paths" envconfig:"INCLUDED_PATHS"`
	AllowedScopes []string	`yaml:"allowed_scopes" envconfig:"ALLOWED_SCOPES"`
	JwksUrl       string	`yaml:"jwks_url"`
```

abel296 replied:
done!

fabiocicerchia commented on server/jwt/jwt.go:
```diff
@@ -0,0 +1,162 @@
+var co *config.Jwt
+var jwtKeyFetcher *jwk.AutoRefresh
+
+func InitJWT(conifg *config.Jwt) {
```
fixing a typo here, although need to change all the occurrences of it.

```suggestion
func InitJWT(config *config.Jwt) {
```

fabiocicerchia commented on server/jwt/jwt.go:
```diff
@@ -0,0 +1,162 @@
+		co = conifg
+		jwtKeyFetcher = jwk.NewAutoRefresh(co.Context)
+		jwtKeyFetcher.Configure(co.Jwks_url,
+			jwk.WithMinRefreshInterval(15*time.Minute),
```
the magic number `15` should either be defined as a constant or, even better, to be configurable.

abel296 replied:
done!

fabiocicerchia commented on server/jwt/jwt.go:
```diff
@@ -0,0 +1,162 @@
+func InitJWTWithDomainConf(domainConfig config.Configuration) {
+	var jwtUrl string
+	var allowedScopes, includedPaths []string
+	if domainConfig.Jwt.Jwks_url != "" {
+		jwtUrl = domainConfig.Jwt.Jwks_url
+	} else {
+		jwtUrl = config.Config.Jwt.Jwks_url
+	}
+	if domainConfig.Jwt.Allowed_scopes != nil {
+		allowedScopes = domainConfig.Jwt.Allowed_scopes
+	} else {
+		allowedScopes = config.Config.Jwt.Allowed_scopes
+	}
+	if domainConfig.Jwt.Included_paths != nil {
+		includedPaths = domainConfig.Jwt.Included_paths
+	} else {
+		includedPaths = config.Config.Jwt.Included_paths
+	}
```
suggesting of skipping the `else`:

```suggestion
	jwtUrl = config.Config.Jwt.Jwks_url
	if domainConfig.Jwt.Jwks_url != "" {
		jwtUrl = domainConfig.Jwt.Jwks_url		
	}
	allowedScopes = config.Config.Jwt.Allowed_scopes
	if domainConfig.Jwt.Allowed_scopes != nil {
		allowedScopes = domainConfig.Jwt.Allowed_scopes
	}
	includedPaths = config.Config.Jwt.Included_paths
	if domainConfig.Jwt.Included_paths != nil {
		includedPaths = domainConfig.Jwt.Included_paths
	}
```

abel296 replied:
done!

fabiocicerchia commented on server/jwt/jwt_test.go:
```diff
@@ -0,0 +1,212 @@
+	ts := CreateTestServer(t, jsonJWKKeySetSingle, jsonJWKKeySetMultiple)
+	defer ts.Close()
+
+	// Without any token
```
would it be possible to split this test into multiple ones based on the scenario tested?

fabiocicerchia commented on server/jwt/jwt_utils.go:
```diff
@@ -0,0 +1,22 @@
+
+import "strings"
+
+func Contains(s []string, searchterm string) bool {
```
this could be replaced by using `ContainsString` from `utils/slice/slice.go`.

abel296 replied:
done!

fabiocicerchia commented on config/model.go:
```diff
@@ -162,6 +165,21 @@ type DomainSet struct {
 
+// Jwt - Defines the config for the jwt validation.
+type Jwt struct {
+	IncludedPaths []string	`yaml:"included_paths" envconfig:"INCLUDED_PATHS"`
```
I've added a new PR for it: https://github.com/Aireuropa/go-proxy-cache/pull/2

abel296 replied:
Thanks for the PR, could you check the changes requested please?

abel296 replied:
Hi @fabiocicerchia, I've already amended this PR changes directly in the commit. I think it is just an easier and more practical way for avoiding conflicts with the last updates

abel296 commented on server/jwt/jwt_functional_test.go:
```diff
@@ -0,0 +1,301 @@
+	assert.Equal(t, http.StatusOK, rr.Code)
+	assert.Equal(t, "MISS", rr.HeaderMap["X-Go-Proxy-Cache-Status"][0])
+
+	// With JWT complete config in common config (with domain)
```
done!

abel296 commented on server/jwt/jwt.go:
```diff
@@ -0,0 +1,172 @@
+var jwtKeyFetcher *jwk.AutoRefresh
+
+func InitJWT(jwtConfig *config.Jwt) {
+	const refreshIntervalMinutes = 15
```
The autorefresh interval is now configurable, done!

abel296 commented:
Hi @fabiocicerchia , we've added a couple of more changes:
- we've put the jwt initialization at both server and domain level (https://github.com/fabiocicerchia/go-proxy-cache/compare/35c881e815019187be5c26317470d4609f7b11bf..06f0b0f106c46bdb981f0dfba25f7a8d62f7fce0)
- we've made the JWKS URL configurable by domain through an environment variable as well, this being, if it exists, the predominant configuration (https://github.com/fabiocicerchia/go-proxy-cache/compare/06f0b0f106c46bdb981f0dfba25f7a8d62f7fce0..ff0bc17f29bdd8e1a7a27e59d2a947a1f49c6e73) . We also added some documentation about it (https://github.com/fabiocicerchia/go-proxy-cache/compare/ff0bc17f29bdd8e1a7a27e59d2a947a1f49c6e73..0b3995e57de9854c9e51d9e8c0c8250571a43ede)

abel296 commented:
Hi @fabiocicerchia , as I added last updates, could you review it please?

abel296 commented:
Hi @fabiocicerchia , these are the last updates:
- excluded paths instead of included paths in JWT config
- regex compatibility with the excluded paths config for excluding dynamic paths
- improved JWT init and JWT handler
- improved testing nginx dockerfile to avoid compatibility issues with otel_ngx_module.so
- migration from github.com/lestrrat-go/jwx to github.com/lestrrat-go/jwx/v2
- update dependencies

fabiocicerchia commented on .env.dist:
```diff
@@ -142,3 +142,28 @@ CACHE_ALLOWED_STATUSES=200,301,302
+#  - JWT_JWKS_URL enviroment variable
+
+# Time in minutes that takes for JWKS to refresh automatically
+# JWT_REFRESH_INTERVAL=
```
Let's use the same example values from `config.yml.dist`:
```suggestion
# JWT_REFRESH_INTERVAL=15
```

fabiocicerchia commented on .env.dist:
```diff
@@ -142,3 +142,28 @@ CACHE_ALLOWED_STATUSES=200,301,302
+
+# --- JWT
+# A list of space-separated paths.
+# JWT_EXCLUDED_PATHS=
```
Let's use the same example values from `config.yml.dist`:
```suggestion
# JWT_EXCLUDED_PATHS=/
```

fabiocicerchia commented on .env.dist:
```diff
@@ -142,3 +142,28 @@ CACHE_ALLOWED_STATUSES=200,301,302
+# JWT_EXCLUDED_PATHS=
+
+# A list of space-separated scopes to be allowed.
+# JWT_ALLOWED_SCOPES=
```
Let's use the same example values from `config.yml.dist`:
```suggestion
# JWT_ALLOWED_SCOPES=scope1, scope2
```

fabiocicerchia commented on config.yml.dist:
```diff
@@ -306,6 +306,19 @@ tracing:
+#   # A list of space-separated paths.
+#   excluded_paths:
+#     - /
+#   # A list of space-separated scopes to be allowed.
```
space or comma separated?

fabiocicerchia commented on config/model.go:
```diff
@@ -162,6 +166,23 @@ type DomainSet struct {
+// Jwt - Defines the config for the jwt validation.
+type Jwt struct {
+	ExcludedPaths       []string   `yaml:"excluded_paths" envconfig:"JWT_EXCLUDED_PATHS" split_words:"true"`
+	AllowedScopes       []string   `yaml:"allowed_scopes" envconfig:"JWT_ALLOWED_SCOPES" split_words:"true"`
```
`scope1, scope2` will end up in having item 0 `scope1,` and item 1 `scope2`

fabiocicerchia commented on docs/CONFIGURATION.md:
```diff
@@ -353,6 +358,19 @@ tracing:
+#   # A list of space-separated paths.
+#   excluded_paths:
+#     - /
+#   # A list of space-separated scopes to be allowed.
```
need to clarify if it is space or comma separated

fabiocicerchia commented on server/jwt/jwt_utils_test.go:
```diff
@@ -0,0 +1,83 @@
+func TestContains(t *testing.T) {
+
+	v := []string{"a", "b"}
+	res := slice.ContainsString(v, "a")
+	if !res {
+		t.Error("Expected true but got", res)
+	}
```
by using the library `github.com/stretchr/testify/assert` present already in other tests this would be more readable:

```suggestion
	res := slice.ContainsString(v, "a")
	assert.True(res)
```

fabiocicerchia commented on test/full-setup/config.yml:
```diff
@@ -46,13 +46,19 @@ domains:
         health_check:
           scheme: http
+    jwt:
+      excluded_paths:
+        - /
```
I believe that if you set these 2 lines under the block on line 49, it'll be inherited by all the domains, without the need of overriding the same value.

fabiocicerchia commented on test/full-setup/nginx/nginx-lua-1.25.4/ubuntu-22.04/.env.dist:
```diff
@@ -0,0 +1,128 @@
+# NGINX MODULES
```
Why is there the need to embed these nginx-lua files?

fabiocicerchia commented on server/jwt/jwt_functional_test.go:
```diff
@@ -0,0 +1,307 @@
+
+	h.ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusMovedPermanently, rr.Code)
```
it would be useful to assert the `Location` header of this 301 respose.

fabiocicerchia commented on server/jwt/jwt_functional_test.go:
```diff
@@ -0,0 +1,307 @@
+
+	h.ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusMovedPermanently, rr.Code)
```
it would be useful to assert the `Location` header of this 301 respose.

fabiocicerchia commented on server/jwt/jwt_functional_test.go:
```diff
@@ -0,0 +1,307 @@
+
+	h.ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusMovedPermanently, rr.Code)
```
it would be useful to assert the `Location` header of this 301 respose.

fabiocicerchia commented on server/jwt/jwt_functional_test.go:
```diff
@@ -0,0 +1,307 @@
+
+	h.ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusMovedPermanently, rr.Code)
```
is it correct that all the tests with a JWT are going to 301?

fabiocicerchia commented on server/jwt/jwt_test.go:
```diff
@@ -0,0 +1,289 @@
+
+	res := getScopes(token)
+
+	assert.ElementsMatch(t, res, []string{"scope1", "scope2", "scope3"}, "Scopes provided doesn't match")
```
```suggestion
	assert.ElementsMatch(t, res, []string{"scope1", "scope2", "scope3"}, "The provided scopes don't match")
```

fabiocicerchia commented on server/jwt/jwt_test.go:
```diff
@@ -0,0 +1,289 @@
+
+	res := getScopes(token)
+
+	assert.ElementsMatch(t, res, []string{"scope1", "scope2", "scope3"}, "Scopes provided doesn't match")
```
```suggestion
	assert.ElementsMatch(t, res, []string{"scope1", "scope2", "scope3"}, "The provided scopes don't match")
```

fabiocicerchia reviewed (changes requested):
Apologies for not getting back to you sooner, I didn't have much time to focus on it.
I'll try spending some more time to test properly this MR in the next following days.

It would be really good not to push force and have a reviewable commit history, because as of now, it's really impossible to figure out what each change contributed to.

I made some more suggestions, but let me test it first before starting fixing, as I might add some more comments, and don't want to let you do too much back and forth.

Thank you for the patience and the effort for putting together this PR!

fabiocicerchia commented on server/jwt/jwt.go:
```diff
@@ -0,0 +1,115 @@
+		return logJWTErrorAndAbort(w, err)
+	}
+	if err := jwt.Validate(token); err != nil {
+		return logJWTErrorAndAbort(w, err)
```
it would be good to increase the test coverage by adding a test case to cover this scenario as well (ie testing the failure of validation).

fabiocicerchia commented on server/jwt/jwt_test.go:
```diff
@@ -0,0 +1,289 @@
+}
+
+func TestRefreshKeySet(t *testing.T) {
+	t.Skip("To run this test, you should set refreshIntervalDuration (from config.go) to time.Second")
```
Can this test be executed automatically, without any manual changes?

fabiocicerchia commented:
@abel296 please have a look at some suggestions made in https://github.com/fabiocicerchia/go-proxy-cache/pull/208

fabiocicerchia commented:
merged all changes + improvements in https://github.com/fabiocicerchia/go-proxy-cache/pull/208
